### PR TITLE
Support the cross compiling for Apple Silicon

### PR DESCRIPTION
**Pull request by wenbingl**
**Description**: 

Support the cross compiling for Apple Silicon

**Motivation and Context**

To build the macos_arm64 binary on a macOS machine with Intel CPU.

### CODE REVIEW

File: BUILD.md [hunk 1]
```diff
@@ -60,24 +60,23 @@ Open Developer Command Prompt for Visual Studio version you are going to use. Th
...
    ```
    ./build.sh --config RelWithDebInfo --build_shared_lib --parallel --use_xcode
    ```
-* With Xcode 12
-   ```
-   ./build.sh --config RelWithDebInfo --build_shared_lib --parallel --use_xcode \
-              --cmake_extra_defines CMAKE_OSX_ARCHITECTURES=x86_64
-   ```
+While without this flag, the cmake build generator will be Unix makefile by default.
+Also, if you want to try cross compiling for Apple Silicon in a legacy MacOS machine(with Intel CPU), please add the argument --osx_arch arm64, however the unit tests will be skipped due to the incompatible CPU instruction set.
```
Comment (gwang-msft) [inline BUILD.md:78]: nit, Also, if you want to cross compile for Apple Silicon in a intel Mac, please ...?
I don't think intel Mac is legacy yet

File: tools/ci_build/build.py [hunk 2]
```diff
@@ -1702,8 +1702,19 @@ def run_csharp_tests(source_dir, build_dir, use_cuda, use_openvino, use_tensorrt
...
+        return True
+    if args.osx_arch != platform.machine():
+        return True
+    return False
+
+
 def build_protoc_for_host(cmake_path, source_dir, build_dir, args):
-    if (args.arm or args.arm64 or args.enable_windows_store) and (not is_windows() and not args.ios):
+    if (args.arm or args.arm64 or args.enable_windows_store) and (
+            not is_windows() and not is_cross_compiling_on_apple(args)):
```
Comment (gwang-msft) [inline tools/ci_build/build.py:1717]: not (is_windows() or is_cross_compiling_on_apple(args))?

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -1966,10 +1977,18 @@ def main():
...
-        elif is_macOS() and args.use_xcode:
-            cmake_extra_args += ['-G', 'Xcode']
+        elif is_macOS():
+            if args.use_xcode:
+                cmake_extra_args += ['-G', 'Xcode']
+            if not args.ios and not args.android and \
+                    args.osx_arch == 'arm64' and platform.machine() == 'x86_64':
+                if args.test:
+                    log.info(
+                        "Cannot test ARM64 build on X86_64. Will skip test running after build.")
```
Comment (gwang-msft) [inline tools/ci_build/build.py:1987]: Raise error here instead of not run test?

**Comment by gwang-msft** [comment]
This is probably out of context, have you tried cross-build for intel from M1 Mac?

File: BUILD.md (hunk 1 above)
Comment (wenbingl) [reply BUILD.md]: using a more accurate name.

File: tools/ci_build/build.py (hunk 2 above)
Comment (wenbingl) [reply tools/ci_build/build.py]: accepted

File: tools/ci_build/build.py (hunk 3 above)
Comment (wenbingl) [reply tools/ci_build/build.py:1987]: the current practice is log.info, I think log.warning is fair enough.

**Comment by wenbingl** [comment]
> This is probably out of context, have you tried cross-build for intel from M1 Mac?

Yes, it works perfectly, and the unit tests also succeed.

**Review by gwang-msft — APPROVED** [review]

**Pull request merged**